src/librustc_typeck/check/mod.rs

```diff
@@ -1311,9 +1311,40 @@ fn check_union(tcx: TyCtxt<'_>, id: hir::HirId, span: Span) {
     def.destructor(tcx); // force the destructor to be evaluated
     check_representable(tcx, span, def_id);
     check_transparent(tcx, span, def_id);
+    check_union_fields(tcx, span, def_id);
     check_packed(tcx, span, def_id);
 }
 
+/// When the `#![feature(untagged_unions)]` gate is active,
+/// check that the fields of the `union` does not contain fields that need dropping.
+fn check_union_fields(tcx: TyCtxt<'_>, _: Span, item_def_id: DefId) -> bool {
+    // Without the feature we check Copy types only later
+    if !tcx.features().untagged_unions {
+        return true;
+    }
+    let item_type = tcx.type_of(item_def_id);
+    if let ty::Adt(def, substs) = item_type.sty {
+        if def.is_union() {
+            let fields = &def.non_enum_variant().fields;
+            for field in fields {
+                let field_ty = field.ty(tcx, substs);
+                // We are currently checking the type this field came from, so it must be local.
+                let field_span = tcx.hir().span_if_local(field.did).unwrap();
+                let param_env = tcx.param_env(field.did);
+                if field_ty.needs_drop(tcx, param_env) {
```

Comment:
Is `needs_drop` sufficiently conservative that it will say `true` when the answer is unknown?

Reply:
Yes, only a `false` return is 100% accurate (i.e. the type was found to never own values of types that may `Drop`, including checking for `Copy` bounds on type parameters).

If you get `true`, you can't assume anything.

E.g. you could use `!ty.needs_drop(tcx, param_env)` to gate skipping `Drop(place: ty)` in MIR.